src/qibolab/platforms/abstract.py

```diff
@@ -175,58 +175,88 @@ def transpile(self, circuit: Circuit):
-
+        # keep track of gates that were already added to avoid adding them twice
+        already_processed = set()
+        # process circuit gates
+        for moment in circuit.queue.moments:
+            moment_start = sequence.finish
+            for gate in moment:
+                if isinstance(gate, gates.I) or gate is None or gate in already_processed:
+                    pass
+
+                elif isinstance(gate, gates.Z):
+                    qubit = gate.target_qubits[0]
+                    sequence.virtual_z_phases[qubit] += np.pi
+
+                elif isinstance(gate, gates.RZ):
+                    qubit = gate.target_qubits[0]
+                    sequence.virtual_z_phases[qubit] += gate.parameters[0]
+
+                elif isinstance(gate, gates.U3):
+                    qubit = gate.target_qubits[0]
+                    # Transform gate to U3 and add pi/2-pulses
+                    theta, phi, lam = gate.parameters
+                    # apply RZ(lam)
+                    sequence.virtual_z_phases[qubit] += lam
+                    # Fetch pi/2 pulse from calibration
+                    RX90_pulse_1 = self.create_RX90_pulse(
+                        qubit,
+                        start=max(sequence.get_qubit_pulses(qubit).finish, moment_start),
+                        relative_phase=sequence.virtual_z_phases[qubit],
+                    )
+                    # apply RX(pi/2)
+                    sequence.add(RX90_pulse_1)
+                    # apply RZ(theta)
+                    sequence.virtual_z_phases[qubit] += theta
+                    # Fetch pi/2 pulse from calibration
+                    RX90_pulse_2 = self.create_RX90_pulse(
+                        qubit, start=RX90_pulse_1.finish, relative_phase=sequence.virtual_z_phases[qubit] - np.pi
+                    )
+                    # apply RX(-pi/2)
+                    sequence.add(RX90_pulse_2)
+                    # apply RZ(phi)
+                    sequence.virtual_z_phases[qubit] += phi
+
+                elif isinstance(gate, gates.M):
+                    # Add measurement pulse
+                    measurement_start = max(sequence.get_qubit_pulses(*gate.target_qubits).finish, moment_start)
+                    gate.pulses = ()
+                    for qubit in gate.target_qubits:
+                        MZ_pulse = self.create_MZ_pulse(qubit, start=measurement_start)
+                        sequence.add(MZ_pulse)
+                        gate.pulses = (*gate.pulses, MZ_pulse.serial)
+
+                elif isinstance(gate, gates.CZ):
+                    # create CZ pulse sequence with start time = 0
+                    cz_sequence = self.create_CZ_pulse_sequence(gate.qubits)
+
+                    # determine the right start time based on the availability of the qubits involved
+                    cz_qubits = {*cz_sequence.qubits, *gate.qubits}
+                    cz_start = max(sequence.get_qubit_pulses(*cz_qubits).finish, moment_start)
```

Comment:
I am not sure why you are not using the register. It would be much more efficient here to just index in a dictonary.

Reply:
The main reason to try not to have to add it is to avoid redundancy, as the information needed is already contained within the PulseSequence. However, I acknowledge that it is not efficient as it currently is.
I was planning on giving it a try to see if I can make PulseSequence more efficient and if not, then introduce the register solution proposed by Stavros.